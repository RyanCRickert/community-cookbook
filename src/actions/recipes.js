import uuid from "uuid";
import database from "../firebase/firebase";

export const addRecipe = (recipe) => ({
  type: "ADD_RECIPE",
  recipe
});

export const startAddRecipe = (recipeData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      name = "",
      ingredients = [],
      instructions = "",
      cookTime = 0,
      createdAt = 0
  } = recipeData;
    const recipe = { name, instructions, createdAt, ingredients, cookTime };

    return database.ref(`recipes`).push(recipe).then((ref) => {
      dispatch(addRecipe({
        id: ref.key,
        ...recipe
      }));
    });
  };
};

export const addIngredient = (id, ingredient) => ({
  type: "ADD_INGREDIENT",
  id,
  ingredient
});

export const startAddIngredient = (id, ingredient) => {
  return (dispatch) => {
    const {
      name = ""
    } = ingredientData;
    const ingredient = {name};

    return database.ref(`recipes/${id}/ingredients`).push(ingredient).then((ref) => {
      dispatch(addIngredient({
        iid: ref.key,
        ingredient
      }));
    });
  };
};

export const editIngredient = (id, iid, update) => {
  type:"EDIT_INGREDIENT",
  id,
  iid,
  update
};

export const startEditIngredient = (id, iid, update) => {
  return (dispatch) => {
    return database.ref(`recipes/${id}/ingredients${iid}`)
  };
};

export const removeRecipe = ({ id } = {}) => ({
  type: "REMOVE_RECIPE",
  id
});

export const startRemoveRecipe = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`recipes/${id}`).remove().then(() => {
      dispatch(removeRecipe({ id }));
    });
  };
};

export const editRecipe = (id, updates) => ({
  type: "EDIT_RECIPE",
  id,
  updates
});

export const startEditRecipe = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`recipes/${id}`).update(updates).then(() => {
      dispatch(editRecipe(id, updates));
    });
  };
};

export const setRecipes = (recipes) => ({
  type: "SET_RECIPES",
  recipes
});

export const startSetRecipes = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid

    return database.ref(`recipes`).once("value", (snapshot) => {
      const recipes = [];

      snapshot.forEach((childSnapshot) => {
        recipes.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setRecipes(recipes));
    });
  };
};

export const setIngredients = (id, ingredients) => ({
  type: "SET_INGREDIENTS",
  id,
  ingredients
});

export const startSetIngredients = () => {
  return (dispatch) => {
    return database.ref(`recipes/${id}/ingredients`).once("value", (snapshot) => {
      const ingredients = [];

      snapshot.forEach((childSnapshot) => {
        recipes.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setIngredients(ingredients));
    });
  };
};