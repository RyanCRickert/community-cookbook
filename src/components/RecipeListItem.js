import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import { firebase } from "../firebase/firebase";

export default class RecipeListItem extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      loggedIn: null
    }
  }

  componentDidMount() {
    firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        this.setState(() => ({ loggedIn: true }))
      } else {
        this.setState(() => ({ loggedIn: false }))
      }
    })
  }

  render() {
    return (
      <Link className="list-item" to={"/" + (this.state.loggedIn ? "edit" : "view") + "/" + this.props.id}>
        <div>
          <h3 className="list-item__title">{this.props.name}</h3>
          <span className="list-item__sub-title">Date added: {moment(this.props.createdAt).format("MMMM Do YYYY")}</span>
        </div>
        <div>
          <h3 className="list-item__data show-for-larger">{this.props.cookTime} {this.props.cookTime === 1 ? "minute" : "minutes"}</h3>
          <span className="list-item__sub-title-right">Feeds: {this.props.feeds}</span>
          <h3 className="list-item__data show-for-mobile">Cooking time: {this.props.cookTime}</h3>
        </div>
      </Link>
    )
  }


}