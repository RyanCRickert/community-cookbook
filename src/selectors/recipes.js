import moment from "moment";

export default (recipes, {name, sortBy, startDate, endDate}) => {
    return recipes.filter((recipe) => {
        const createdAtMoment = moment(recipe.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true ;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true ;
        const nameMatch = recipe.name.toLowerCase().includes(name.toLowerCase());

        return startDateMatch && endDateMatch && nameMatch;
    }).sort((a , b) => {
        if (sortBy === "date") {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === "name") {
            return a.name > b.name ? 1 : -1;
        } else if (sortBy === "cookTime") {
            return a.cookTime < b.cookTime ? 1 : -1;
        } else if (sortBy === "feeds") {
            return a.feeds < b.feeds ? 1 : -1;
        }
    });
};