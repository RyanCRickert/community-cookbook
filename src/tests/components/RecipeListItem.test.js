import React from "react";
import { shallow } from "enzyme";
import RecipeListItem from "../../components/RecipeListItem";
import recipes from "../fixtures/recipes";

test("should render RecipeListItem correctly", () => {
  const wrapper = shallow(<RecipeListItem {...recipes[0]}/>)
  expect(wrapper).toMatchSnapshot();
});