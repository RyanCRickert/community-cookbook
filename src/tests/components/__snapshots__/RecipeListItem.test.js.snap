// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RecipeListItem correctly 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/view/0"
>
  <div>
    <h3
      className="list-item__title"
    >
      Baked Chicken
    </h3>
    <span
      className="list-item__sub-title"
    >
      Category: 
      Chicken
    </span>
  </div>
  <div>
    <h3
      className="list-item__data show-for-larger"
    >
      45
       
      minutes
    </h3>
    <span
      className="list-item__sub-title-right"
    >
      Feeds: 
      4
    </span>
    <h3
      className="list-item__data show-for-mobile"
    >
      Cooking time: 
      45
       
      minutes
    </h3>
  </div>
</Link>
`;
